File: src/main/python/AoC2015_11.py
#! /usr/bin/env python3
#
# Advent of Code 2015 Day 11
#

import re
import sys

from aoc.common import InputData
from aoc.common import SolutionBase
from aoc.common import aoc_samples

ALPH = "abcdefghijklmnopqrstuvwxyz"
NEXT = "bcdefghjjkmmnppqrstuvwxyza"
CONFUSING_LETTERS = {"i", "o", "l"}
RE = re.compile(r"([a-z])\1")

Input = str
Output1 = str
Output2 = str


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        return next(iter(input_data))

    def is_ok(self, password: str) -> bool:
        if len(set(RE.findall(password))) < 2:
            return False
        trio = False
        for i in range(len(password)):
            if password[i] in CONFUSING_LETTERS:
                return False
            trio = trio or (
                i < len(password) - 3
                and ord(password[i]) == ord(password[i + 1]) - 1
                and ord(password[i + 1]) == ord(password[i + 2]) - 1
            )
        return trio

    def get_next(self, password: str) -> str:
        def increment(password: str, i: int) -> str:
            password = (
                password[:i]
                + NEXT[ALPH.index(password[i])]
                + password[i + 1 :]
            )
            if password[i] == "a":
                password = increment(password, i - 1)
            return password

        while True:
            password = increment(password, len(password) - 1)
            if self.is_ok(password):
                break
        return password

    def part_1(self, password: Input) -> Output1:
        return self.get_next(password)

    def part_2(self, password: Input) -> Output2:
        return self.get_next(self.get_next(password))

    @aoc_samples(
        (
            ("part_1", "abcdefgh", "abcdffaa"),
            ("part_1", "ghijklmn", "ghjaabcc"),
        )
    )
    def samples(self) -> None:
        assert not self.is_ok("abci")
        assert not self.is_ok("hijklmmn")
        assert not self.is_ok("abbceffg")
        assert not self.is_ok("abbcegjk")


solution = Solution(2015, 11)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()

File: src/main/java/AoC2015_11.java
import static com.github.pareronia.aoc.StringOps.nextLetter;

import java.util.List;

import com.github.pareronia.aoc.solution.Sample;
import com.github.pareronia.aoc.solution.Samples;
import com.github.pareronia.aoc.solution.SolutionBase;

public class AoC2015_11 extends SolutionBase<String, String, String> {
    
    private AoC2015_11(final boolean debug) {
        super(debug);
    }

    public static final AoC2015_11 create() {
        return new AoC2015_11(false);
    }

    public static final AoC2015_11 createDebug() {
        return new AoC2015_11(true);
    }
    
    @Override
    protected String parseInput(final List<String> inputs) {
        assert inputs.size() == 1;
        return inputs.get(0);
    }

    @Override
    public String solvePart1(final String input) {
        return PasswordGenerator.generateFrom(input);
    }

    @Override
    public String solvePart2(final String input) {
        return PasswordGenerator.generateFrom(PasswordGenerator.generateFrom(input));
    }

    @Override
    @Samples({
        @Sample(method = "part1", input = "abcdefgh", expected = "abcdffaa"),
        @Sample(method = "part1", input = "ghijklmn", expected = "ghjaabcc"),
        @Sample(method = "part1", input = "heqaabcc", expected = "heqbbcdd"),
    })
    public void samples() {
    }

    public static void main(final String[] args) throws Exception {
        assert PasswordChecker.isOk("heqaabcc".toCharArray()) == true;

        AoC2015_11.create().run();
    }
    
    private static final class PasswordChecker {
        
        public static boolean isOk(final char[] password) {
            final int[] pairs = new int[] { -1, -1 };
            boolean trio = false;
            for (int i = 0; i < password.length; i++) {
                if (password[i] == 'i' || password[i] == 'o' || password[i] == 'l') {
                    return false;
                }
                if (i > 0 && password[i] == password[i - 1]) {
                    pairs[pairs[0] == -1 ? 0 : 1] = i;
                }
                trio = trio
                        || i < password.length - 3
                        && password[i] == password[i + 1] - 1
                        && password[i + 1] == password[i + 2] - 1;
            }
            return trio && pairs[0] != -1 && pairs[1] != -1
                    && password[pairs[0]] != password[pairs[1]];
        }
    }
    
    private static final class PasswordGenerator {
        
        private static char[] increment(final char[] password, final int i) {
            password[i] = nextLetter(password[i], 1);
            if (password[i] == 'a') {
                increment(password, i - 1);
            }
            return password;
        }
        
        public static String generateFrom(final String password) {
            char [] chars = password.toCharArray();
            do {
                chars = increment(chars, chars.length - 1);
            } while (!PasswordChecker.isOk(chars));
            return String.valueOf(chars);
        }
    }
}
